feat: adding `--tag` in the bsf oci command

hanshal101 opened the pull request:
CLOSES #38 
FIXES #38 

## Description
This PR adds the feature to mention the tag in the oci command.

dr-housemd commented:
@hanshal101 the tag is on this field. Right now, users can pass `ttl.sh/app:v0.0.1` in the HCL file. 
We want them to pass `ttl.sh/app` in the HCL file and `--tag=v0.0.1` from CLI. 
Considering that, we have to read the CLI flag i.e `v0.0.1` and append it to the name.  Hope that makes sense
 https://github.com/buildsafedev/bsf/blob/main/pkg/hcl2nix/export.go#L12

dr-housemd commented:
@hanshal101  yes, that would be a step in the right direction. 

If `artifact.Name` contains a `:` character, we need to replace it with what user passed in `tag` .

hanshal101 commented:
Changes made 👍

dr-housemd commented on cmd/oci/oci.go:
```diff
@@ -61,6 +62,10 @@ var OCICmd = &cobra.Command{
 		platform = p
 
+		if tag != "" {
+			artifact.Name = fmt.Sprintf("%s:%s", artifact.Name, tag)
```
if `artifact.Name` contains a `:` (ex: a user passed `ttl.sh/app:1h` and they want to override the tag with `3h`), we should allow that.

rakshitgondwal commented on cmd/oci/oci.go:
```diff
@@ -61,6 +62,15 @@ var OCICmd = &cobra.Command{
+		if tag != "" {
+			if strings.Contains(artifact.Name, ":") {
+				parts := strings.Split(artifact.Name, ":")
+				artifact.Name = fmt.Sprintf("%s:%s", parts[0], tag)
```
Changing the `artifact.Name` here will only affect the name when we are loading the artifact to docker daemon.

This is the bit where we are passing the name to the flake generation:
https://github.com/buildsafedev/bsf/blob/0cb31b1050f2622da13663f1206b545b5b3241dd/pkg/nix/template/oci.go#L29

and this flake is being generated from here:
https://github.com/buildsafedev/bsf/blob/0cb31b1050f2622da13663f1206b545b5b3241dd/pkg/generate/generate.go#L20

In this same function, you will see we fetch the config again by reading `bsf.hcl` which means the name will remain the same for the artifact.

What we can do is, just read the config in oci command itself and then pass it further to `generate.Generate` function over here:
https://github.com/buildsafedev/bsf/blob/0cb31b1050f2622da13663f1206b545b5b3241dd/cmd/oci/oci.go#L69
so that we have the tag in the nix flake as well.

hanshal101 commented:
Changes Done

hanshal101 commented:
Hey @rakshitgondwal I have implemented the `map` method discussed, just have a look at it also I have added one more step of checking the `len` of the map at generate function if its greater than 0, then the loop runs.

rakshitgondwal commented on pkg/generate/generate.go:
```diff
@@ -29,6 +29,14 @@ func Generate(fh *hcl2nix.FileHandlers, sc buildsafev1.SearchServiceClient) erro
 		return fmt.Errorf("%v", &dstErr)
 	}
 
+	if len(nameMap) > 0 {
+		for _, artifact := range conf.OCIArtifact {
+			if newName, ok := nameMap[artifact.Name]; ok {
+				artifact.Name = newName
+			}
+		}
+	}
```
```suggestion
	if len(nameMap) > 0 {
		for i, artifact := range conf.OCIArtifact {
			if newName, ok := nameMap[artifact.Name]; ok {
				conf.OCIArtifact[i].Name = newName
			}
		}
	}
```

rakshitgondwal commented on cmd/oci/oci.go:
```diff
@@ -60,13 +61,26 @@ var OCICmd = &cobra.Command{
 		platform = p
+		nameMap := make(map[string]string)
+		originalName := artifact.Name
+		newName := artifact.Name
```
```suggestion
		var newName string
```

rakshitgondwal commented on cmd/oci/oci.go:
```diff
@@ -60,13 +61,26 @@ var OCICmd = &cobra.Command{
+		if tag != "" {
+			if strings.Contains(artifact.Name, ":") {
+				parts := strings.Split(artifact.Name, ":")
+				newName = fmt.Sprintf("%s:%s", parts[0], tag)
```
Can you check for the length of `parts` to avoid any panic errors?

hanshal101 commented:
Also in the codebase I'm getting some warnings should I fix them?
Some are like this:
![image](https://github.com/buildsafedev/bsf/assets/122217807/e8d1ace3-11e7-4ac5-a8a8-6682da7f09af)

dr-housemd commented:
> Also in the codebase I'm getting some warnings should I fix them?

We would like them fixed but likely as a different PR. 
We also want to prevent them from happening again so we need a CI job to run the linter and fail when we see it.

dr-housemd commented on cmd/oci/oci.go:
```diff
@@ -60,13 +61,30 @@ var OCICmd = &cobra.Command{
 			os.Exit(1)
 		}
 		platform = p
+		nameMap := make(map[string]string)
```
better variable name please - it should be related to the domain/action.

hanshal101 replied:
how does `articfactName` sounds?

dr-housemd replied:
sounds good

dr-housemd commented on cmd/oci/oci.go:
```diff
@@ -60,13 +61,30 @@ var OCICmd = &cobra.Command{
+		originalName := artifact.Name
+		var newName string
+
+		if tag != "" {
```
If a user did not pass the `tag` flag and `artifact.Name` doesn't include a `:`, we should append `:latest` tag.

dr-housemd commented on pkg/generate/generate.go:
```diff
@@ -17,7 +17,7 @@ import (
 
 // Generate reads bsf.hcl, resolves dependencies and generates bsf.lock, bsf/flake.nix, bsf/default.nix, etc.
-func Generate(fh *hcl2nix.FileHandlers, sc buildsafev1.SearchServiceClient) error {
+func Generate(fh *hcl2nix.FileHandlers, sc buildsafev1.SearchServiceClient, nameMap map[string]string) error {
```
I would suggest to not modify the Generate function's signature for this case. 

We should expect the `conf.OCIArtifact[i].Name` to be modified before it is passed here. cc : @rakshitgondwal
